### ISSUE TITLE
JMAP: Better broken MIME charset handling

### ISSUE THREAD

**Issue by robmueller** [cyrusimap/cyrus-imapd#2471]
Unfortunately real world emails often don't follow the standards as nicely as one would hope. The FastMail middleware has a bunch of hacks in it to try and deal with these that we now need to push down into cyrus.

The two main ones I'd like to see:

1. MIME encoded headers where a character is broken between multiple MIME words. This is clearly a spec violation, but seen in the wild.
```
  Each 'encoded-word' MUST represent an integral number of characters.
  A multi-octet character may not be split across adjacent 'encoded-
  word's.
```
I think the simple solution to this is if two mime words next to each other separated by only whitespace have the same charset + same encoding, then we should just concat the bytes together before decoding. This would fix all the cases of this that I've seen.

2. Message bodies with bogus character set encodings or no specified encoding at all. We've seen this semi-regularly with various japanese emails and also hebrew emails.

Handling this is a heuristic mess, but a first pass solution that would probably catch 90% of cases:

a) If the charset is us-ascii (or not present), and there's octets outside the [\r\n\t\x20-\x7e] range (e.g. any high bit, or control chars like ESC), then run the mozilla charset detector over it and use that if it detects one instead
b) If the charset is utf-8, and decoding fails with many errors, then run the mozilla charset detector over it and use that if it detects one instead

https://github.com/Joungkyun/libchardet

**Comment by robmueller** [comment]
Ok, one more I'd like to see, we have a hack to deal with a few emails we saw with UTF-32 but with the wrong endianness. In theory "unmarked form uses big-endian byte serialization by default", but we've seen the reverse :(

The current solution is to look at 4 chars (16 bytes), and if any of them are > 0x10ffff (e.g. invalid unicode code points), then we assume it's actually LE instead.

  my @chars = unpack('N*', substr($_[0], 0, 16)); # 4 chars
  my $hichars = grep { $_ > 0x10ffff } @chars;
  unshift @_, "utf-32" . ($hichars ? "le" : "be");

No idea if this is hard to implement in the charset decoder or not...

**Comment by rsto** [comment]
Point 1 (multi-octet characters across MIME header words) is fixed by https://github.com/cyrusimap/cyrus-imapd/commit/7c528574bd468550495499bfcfb04b43ddb70417

**Comment by rsto** [comment]
For UTF-32 bodies I've just pushed https://github.com/cyrusimap/cyrus-imapd/commit/21739c7d3dd138afd446e8c20b8e5788f7e2a471 (see its related test https://github.com/cyrusimap/cassandane/commit/e9a74a5ae73f738aaa551c28a3bc7dfaad5728f3).

I've experimented with the charset detection that's built in ICU (the charset library that backs the Cyrus charset API), but found it to produce subpar results. There's at least https://github.com/Joungkyun/libchardet (Mozilla) and https://github.com/google/compact_enc_det (Google) to evaluate for use in Cyrus. The Mozilla library looks to be plain C vs Google's C++ library, so it has a slight advantage.
